staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/upgrade.go

```diff
@@ -32,6 +34,19 @@ const HeaderSpdy31 = "SPDY/3.1"
 type responseUpgrader struct {
 }
 
+// connWrapper is used to wrap a hijacked connection its bufio.Reader. All
+// calls will be handled directly by the underlying net.Conn with the exception
+// of Read calls, which will read from the bufio.Reader. This ensures that data
+// already inside the used bufio.Reader instance is also read.
+type connWrapper struct {
+	net.Conn
+	bufReader *bufio.Reader
+}
+
+func (w *connWrapper) Read(b []byte) (n int, err error) {
```

Comment:
This solution is okay, but has one edge case. When Close is called, it is entirely possible for Read to still return with data instead of EOF. That may or may not be surprising behavior.

To fix this, you could discard the buffer upon Close.
```go
func (w *connWrapper) Close() error {
    w.bufReader.Discard(w.bufReader.Buffered())
    return w.Conn.Close()
}
```